src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,165 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final Object[] EMPTY_DATAARRAY = {};
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] dataArray;
+    private int size;
+
+    public ArrayList() {
+        this.dataArray = new Object[DEFAULT_CAPACITY];
+    }
+
+    public ArrayList(int capacity) {
+        if (capacity == 0) {
+            this.dataArray = EMPTY_DATAARRAY;
+        } else if (capacity < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Capacity can't be negative");
+        } else {
+            this.dataArray = new Object[capacity];
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        growIfFull();
+        addOnNextFreeSell(value);
     }
 
     @Override
     public void add(T value, int index) {
-
+        indexValidation(index);
+        if (index == size && size != dataArray.length) {
+            addOnNextFreeSell(value);
+        } else {
+            addByIndex(value, index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list.isEmpty()) {
+            return;
+        }
+        if (list.size() + size > dataArray.length) {
+            growTo(dataArray, list);
+        }
+        for (int i = 0; i < list.size(); i++) {
 
+            this.add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
+        if (isIndexExist(index)) {
+            return (T) dataArray[index];
+        }
         return null;
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (isIndexExist(index)) {
+            dataArray[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
+        if (isIndexExist(index)) {
```

Comment:
this is why I do not recommend to mix up the boolean and exception validation approaches `if` here is essentially redundant because there will never be a situation when it is false, because it will just throw an exception